File: src/main/java/com/github/eendroroy/sdk/openweather/converter/ResponseConverter.java
package com.github.eendroroy.sdk.openweather.converter;

import com.github.eendroroy.sdk.openweather.response.GetWeatherResponse;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Response;
import retrofit2.Retrofit;

import java.io.IOException;
import java.lang.annotation.Annotation;

/**
 * @author indrajit
 */

public class ResponseConverter<T> {
    Retrofit retrofit;
    Response<T> response;

    public ResponseConverter(Retrofit retrofit, Response<T> response){
        this.retrofit = retrofit;
        this.response = response;
    }

    public T convert() throws IOException {
        if(response.isSuccessful()) {
            return response.body();
        } else {
            Converter<ResponseBody, GetWeatherResponse> errorConverter =
                    retrofit.responseBodyConverter(GetWeatherResponse.class, new Annotation[0]);
            assert response.errorBody() != null;
            return (T) errorConverter.convert(response.errorBody());
        }
    }

}

File: src/main/kotlin/com/github/eendroroy/sdk/openweathermap/converter/ResponseConverter.kt
package com.github.eendroroy.sdk.openweathermap.converter

import com.github.eendroroy.sdk.openweathermap.response.BaseResponse
import retrofit2.Response
import retrofit2.Retrofit
import java.io.IOException

/**
 * @author indrajit
 */
class ResponseConverter<T>(var retrofit: Retrofit, var apiResponse: Response<T>) {

    @Throws(IOException::class)
    inline fun <reified T> convert(): T? {
        val response: T? = if (apiResponse.isSuccessful) {
            apiResponse.body() as T?
        } else {
            val errorConverter = retrofit.responseBodyConverter<T>(T::class.java, arrayOfNulls(0))
            errorConverter.convert(apiResponse.errorBody()!!)
        }

        if (response is BaseResponse) {
            (response as BaseResponse).httpCode = apiResponse.code()
            (response as BaseResponse).httpMessage = apiResponse.message()
        }

        return response
    }
}